README.md
# 1 Introduction

## 1.1 What is FPL?

FPL, the Formal Proving Language, is a specification for an artificial language 
that is going to serve as a universal language to formulate
mathematical definitions, theorems, and proofs independently of 
local natural languages. 

Watch the video:

[![Watch the video](https://img.youtube.com/vi/ROCNCDRiBp0/0.jpg)](https://www.youtube.com/watch?v=ROCNCDRiBp0)

While English is currently the de facto standard natural language for
mathematical publications, FPL should further facilitate the dialog between 
mathematicians worldwide, providing a standard for notation, and the  
degree of details of mathematical proofs, definitions, axioms, and theorems. 

At the same time, as a formal language, FPL aims to facilitate 
the automation of:
* parsing and translating the code written in FPL into an arbitrary human 
local language (as a bridge between FPL, which is a formal language and 
any natural language),
* verifying, if a code written in FPL (a mathematical theory) is logically
consistent (i.e. without contradictions), 
* helping to find a proof for a theorem, given a theory written 
in FPL,   
* generating theories based on axioms written in FPL,

to mention only some possible applications.

## 1.2 Why FPL?

FPL is not yet another automated proof assistant or yet another automated theorem proving tool. It attempts to set a standard for writing mathematical definitions, theorems, and proofs while encouraging automated programming tools based on FPL. 

The syntax of existing formal systems and automated tools is computer-friendly, i.e., primarily designed to be interpreted by computers, not humans. Therefore, the language used by these tools is hard to read and to understand, even by mathematicians. The communities using these tools are small communities of specialists. Despite that, such automated tools provide many advantages. In particular, they help to write correct mathematical proofs and help to avoid errors. 

Therefore, it would be great to use a formal language that is readable, catchy, and easy to learn, even for non-programmers. Such tools could be (ideally) also used for educational purposes to teach mathematics. 

FPL wants to close this gap. FPL will have a human-friendly syntax while being formal enough to develop tools based on it that can assist in writing and developing correct mathematical theories. In particular, FPL should be easy enough to be used in mathematical education. 

Thus, FPL primarily aims at being human-readable and only then at being compiled and processed by computers. FPL should enjoy better user acceptance by this human-centric approach while providing comparable advantages that automated formal language tools provide.

# 2 High-Level Conceptual Design

The project to develop FPL started June 14, 2020, as described in at [https://www.bookofproofs.org/branches/fpl-formal-proving-language/](https://www.bookofproofs.org/branches/fpl-formal-proving-language/).
Originally, developing the specification was planned as a collaborative project. Unfortunately, and probably due to the pandemic,
no community formed to work on it. Meanwhile, I could work on the project I initiated on my own. 
I identified the following High-Level Requirements for the FPL language:

| No.        | Requirement for FPL         | Level  |
| :-------------: |-------------| :-----:|
|1 |Express proof-based mathematics (PBM) only, nothing else.|MUST|
|2 |Support a clear separation of the concepts of truth and accuracy.|MUST|
|3 |Express PBM in a structured code to distinguish it from text passages written in prose.|MUST|
|4 |Support eight building blocks of PBM: definitions, theorems, propositions, lemmas, corollaries, axioms, proofs, and conjectures.|MUST|
|5 |Allow embedding FPL code into surrounding text that might contain non-PBM contents.|MUST|
|6 |Support localization|SHOULD|
|7 |Allow distinguishing four theorem-like building blocks: theorems, propositions, lemmas, and corollaries.|SHOULD|
|8 |Allow zero to many proofs of theorem-like building blocks and distinguish them syntactically from unproven conjectures.|MUST|
|9 |Disambiguate FPL code on a syntactical level using appropriate parentheses rules.|MUST|
|10 |Allow only deterministic interpreters to disambiguate FPL code on a semantical level.|MUST|
|11 |Incorporate the syntax of predicate logic (at least PL2).|MUST|
|12 |Allow referring to (unproven) conjectures in mathematical proofs and interpret such proofs accordingly.|MUST|
|13 |Support nesting and linking different logical steps and proofs into more complex ones.|MUST|
|14 |Use definitions as a meta-language to introduce new syntax and new domains of discourses allowing to interpret the formulas.|MUST|
|15 |Definitions introduce new types that can be used to declare FPL variables.|MUST|
|16 |Support for asserting and checking a type of a variable (`is` operator)|MUST|
|17 |Support flexible notation while defining new predicates, functional terms, or mathematical objects.|MUST|
|18 |Allow a free configuration of axioms and inference rules.|MUST|
|19 |Look&feel of notation, possibly similar modern notation. Ideally, exploit the possibilities of LaTeX.|SHOULD| 
|20 |Accept different levels of detail in mathematical proofs. If requested, provide additional details for validated proofs.|SHOULD|
|21 |Independence from foundations: No built-in axiomatic system of PBM in the syntax of FPL.|MUST|
|22 |Support both intuitionistic and non-intuitionistic arguments and mathematical objects.|SHOULD|
|23|Support definitions with compound parameters having implicit properties.|MUST|
|24|Support for relative definitions|MUST|
|25|Support for intrinsic definitions|MUST|
|26|Syntax independent from notation|MUST|
|27|Allow ranges of variables, including *countable* or *uncountable*, *ordered* or *unordered*, *finite*, or *infinite*.|SHOULD|
|28|Support inheritance and overriding of properties of parent classes when defining new *types*.|MUST|
|29|Determining the type (`is` operator) reflects the inheritance tree.|MUST|
|30|Stating axioms inside a definition as mandatory properties or as separate building blocks.|SHOULD|
|31|Support of definitions of functional terms|MUST|
|32|Support of definitions of predicates|MUST|
|33|Clear scope of variables in which they are declared.|MUST|
|34|Support to declare variables in the defined types, including support for at least PL2|MUST|
|35|Support assignment of values or expressions to variables|MUST|
|36|Support to delegate the interpretation of intrinsic definitions|MUST|
|37|FPL to allow identification by providing means of casting different data types to each other.|SHOULD|
|38|Allow recursive linguistic constructs|MUST|
|39|Allow loops|MUST|
|40|Allow self-reference in definitions.|MUST|
|41|Support generic types.|MUST| 
|42|Check the self-containment (FPL interpreter)|MUST|

When I started identifying these requirements, I quickly realized that it would not suffice to publish them and create a proof of concept. During my analysis, it turned out that many expert views on this subject may deviate from what I'm proposing. Therefore, I decided to publish my analysis in a 
separate paper that *justifies* all these requirements. You can find the paper at ["Formal Proving Language (FPL)
A Proposal How To Write and Read Proof-Based Mathematics (PBM)"](https://www.bookofproofs.org/FPLHighLevelDesign.pdf).

The paper also contains a description of a proof of concept that is based on some first FPL code examples you can
find in this repository as well as the syntax diagrams of the FPL grammar (version 1.0.0).

# 3 Proof of Concept (Current State)

There is a proof of concept (POC) based on the above requirements. It currently consists of the following pieces:

1. The FPL grammar in an EBNF. 
  * [Current syntax diagrams](https://www.bookofproofs.org/FPLSyntaxDiagrams.xhtml)
  * [EBNF definition](https://github.com/bookofproofs/fpl/tree/master/grammar/)
2. [FPL demo theories ](https://github.com/bookofproofs/fpl/tree/master/poc/theories/)
3. A demo python project consisting of a parser generated using the [TatSu package](https://tatsu.readthedocs.io/en/stable/).

# 4 Contributing to the Project

The project is still ongoing, and you are invited to collaborate. Please refer to the [contributing section](https://github.com/bookofproofs/fpl/blob/master/CONTRIBUTING.md) for details.

# 5 Getting Started

## Software dependencies
* Use python (tested with 3.10 or higher).
* Install the following packages:
  * pytest (7.1.3 or higher) 
  * parameterized (0.8.1 or higher)
  * tatsu (5.8.3 or higher) 
  * anytree (2.8.0 or higher)
  * z3-solver (4.11.2 or higher)
## Testing
* Run the [poc tests](https://github.com/bookofproofs/fpl/tree/master/poc/tests) using pytest. (Note: In many python IDEs, there is a shortcut for doing it. For instance, in PyCharm you can just right-click the fpl/poc/tests folder and select 'Run pytest in tests'). 
* If the tests are runnable, your repository is all set.
## Trying out the IDE
* Run fpl/ide/fplide.py and open one of the FPL files in the folder [fpl/poc/theories](https://github.com/bookofproofs/fpl/tree/master/poc/theories).







ide/CHANGES.md
# Changes in the FPL IDE
## 1.6.14
* Feature
  * Adjustments to reflect the refactoring work done in FPL interpreter version 1.9.0 
## 1.6.13
* Feature:
  * Self-containment info box added (yet empty)
## 1.6.12
* Bugfixes:
  * Run-time errors will be displayed in the error list and do not cause the program to halt.
## 1.6.11
* Bugfixes:
  * Number of auto-tabs in the new line after pressing <Return> corrected.
  * Word-wrap for long error messages introduced in 1.6.10 removed since it did not reliable enough.
* Refactoring:
  * A separate new class for handling the InfoBoxes.py. Code separated from the main code in the fplide.py file.
## 1.6.10
* Redesign:
  * "Rebuild" menu item renamed to "Verify"
* Feature:
  * Word-wrap for long error messages 
  * More meaningful syntax errors
* Bugfixes:
  * Missing message and position for syntax errors do appear now in the error list and can be double-clicked to jump to the position in the code.
  * Runtime error when re-verifying a file with a yet another syntax error is corrected  
## 1.6.9
* Bugfixes
  * The key shortcuts Ctrl + x changed to Ctrl + q and Ctrl + S to Ctr + s to be more intuitive 
  * Wrong relative target path when saving files
  * Attribute error when saving an opened file
## 1.6.8
* Feature
  * Highlighting of PredicateIdentifiers added
* Bugfixes
  * The keywords 'premise', 'undefined', and 'template' wer not highlighted properly
## 1.6.7
* Bugfixes
  * The wait cursor was not changing during the busy status build process
## 1.6.6
* Refactoring
  * The syntax and semantical analysis and function was moved from FrameWithLineNumbers object to FplIde object, because it has to be done not for a single opened FPL file but for the whole FPL theory that might consist of many FPL files.
  * Consequently, the bindings Alt-Control-g and Alt-Control-l were deprecated.
* Bugfixes
  * Syntax highlighting was not available for files of the theory that were opened by the user in addition to the main FPL theory file.
## 1.6.5
* Bugfixes
  * Build current FPL discontinued since building a single file in the context of a whole theory is not feasible with respect to context-specific errors: we use build whole theory instead.
  * AttributeError save_file when building theory fixed.
  * Double-click on default constructors led to a run-time error since there was no position to jump to. Those default constructors are not in the source code and the objects are defined intrinsically.
* Refactoring
  * One central static method for writing files to disk in futils.py
## 1.6.4
* Bugfixes:
  * File names in error list are now error-location specific
  * Error-positions from other files were highlighted also in the currently opened file
  * Double-clicking the error of a file that was not opened did not work  
  * Highlighting of variables was broken
  * Affected symbol table test cases adjusted to the new positions
## 1.6.3
* Bugfixes:
  * no more duplicates of variable-related errors and warnings
* Feature:
  * Errors and warnings are now sorted by severity and occurance (file / line / column)  
## 1.6.2
* Refactoring:
  * A separate new class for the ObjectBrowser.py. Code separated from the main code in the fplide.py file.
  * Name column removed from the ObjectBrowser
* Bugfixes:
  * Wait cursor added
  * Namespaces sorted alphabetically
  * File names removed from object browser for pure outline lines that do not refer to actual source code. 
  * Event "double-click" instead of "click" bound to the Error List
* Feature:
  * Double-clicking the object browser will open a file of the theory and jump to the respective position in the source code.
## 1.6.1
* Bugfixes:
  * AttributeError: 'FplIde' object has no attribute 'fpl_source_transformer' when reformatting code corrected. 
  * Higlighting and (any previous) positions of errors in the FPL code are now correctly calculated after reformatting the code. 
  * When saving a file, the cursor stays at the last position in the editor end does not jump to its beginning. 
  * When closing a theory, the symbol table gets cleared so that it does not mix up with a new theory to be opened.   
## 1.6.0
* Feature: 
  * Command Build->Current FPL File - when there are FPL interpreter errors in an FPL file and the user corrects them in the editor, the menu command will re-interpret the file again and creates an amended error list.  
* Refactoring
  * Menu "File" deprecated (removed)
* Minor bug-fixes
## 1.5.0
* Feature: 
  * A user dialog for the menu command FPL Theory->Open
  * Command FPL Theory->Close
## 1.4.2
* Bugfix: Menu command Fpl Theory->New now corrects the disabled status of the menu after creating a new theory.
* Refactoring: New class IdeModel to store the state of the ide and main data or handler pointers.
## 1.4.1
* Bugfix: Menu command Fpl Theory->New now opens the FPL file after creating it.
## 1.4.0
* Feature: A user dialog for the menu command FPL Theory->New.
* Refactoring: 
  * A class for common dialogs, 
  * SettingsDialog renamed in DialogSettings
## 1.3.1
* Refactoring and re-design of the menu items (without implementation).
* Help guide file markdown added.
## 1.3.0
* Features: The object tree visualizes the structure of the mathematical theory opened in the IDE.
## 1.2.9
* Features: 
  * The error list now shows the diagnosis id of errors instead of the class name of the error
  * Object explorer shows an example of how it will be supposed to look like later.
  * A more readable font family in Error list
  * Tabs "Syntax Browser" and "Semantic Browser" discontinued
  * Tab "Output" renamed in "Debug"
* Bug fixes:
  * Each editor cursor column in the status bar starts at 1 (instead of 0)
  * Double-clicking an error from error list jumps to the correct starting column (instead of starting colum + 1) in the editor.
  * Errors are red underlined in code editor (preliminary implementation - only the first character)

ide/InfoBoxes.py
import tkinter as tk
from tkinter import ttk, messagebox
from poc.util.fplutil import Utils


class InfoBoxes:
    def __init__(self, panedWindowBelowEditor, ide):
        self.ide = ide
        self.tree_view_errors = None
        self._tabControl = ttk.Notebook(panedWindowBelowEditor)
        self.__add_error_info_box()

        self._frameDebug = ttk.Frame(self._tabControl)
        self._frameDebug.config()

        self._frameSelfContainment = ttk.Frame(self._tabControl)
        self._frameSelfContainment.config()

        self._tabControl.add(self._gridErrors, text='Error List')
        self._tabControl.add(self._frameDebug, text='Debug')
        self._tabControl.add(self._frameSelfContainment, text='Self-Containment')

        self._tabControl.pack(expand=True, fill=tk.BOTH)

    def __add_error_info_box(self):
        self._gridErrors = ttk.Frame(self._tabControl)
        self._gridErrors.pack(expand=True, fill=tk.BOTH)

        # number of errors label
        self._label_error_num = tk.Label(self._gridErrors, text="Errors (0)",
                                         relief=tk.GROOVE, bg=self.ide.theme.get_bg_color(),
                                         fg=self.ide.theme.get_fg_color())
        self._label_error_num.grid()
        self._label_error_num["compound"] = tk.LEFT
        self._label_error_num["image"] = self.ide.model.images["cancel"]
        self._label_error_num.grid(row=0, column=0, sticky=tk.W + tk.E, pady=2)

        # number of warnings label
        self._label_warning_num = tk.Label(self._gridErrors, text="Warnings (0)",
                                           relief=tk.GROOVE, bg=self.ide.theme.get_bg_color(),
                                           fg=self.ide.theme.get_fg_color())
        self._label_warning_num.grid()
        self._label_warning_num["compound"] = tk.LEFT
        self._label_warning_num["image"] = self.ide.model.images["warning"]
        self._label_warning_num.grid(row=0, column=1, sticky=tk.W + tk.E, pady=2)

        # option menue for filtering all or current file errors
        self._all_file_list = ["All", "Current File"]
        self._all_file_option_menu = tk.OptionMenu(self._gridErrors,
                                                   tk.StringVar(self._gridErrors, value="All"),
                                                   *self._all_file_list)
        self._all_file_option_menu.config(relief=tk.GROOVE,
                                          bg=self.ide.theme.get_bg_color(),
                                          highlightbackground=self.ide.theme.get_bg_color(),
                                          fg=self.ide.theme.get_fg_color())
        self._all_file_option_menu.grid(row=0, column=2, sticky=tk.W + tk.E, pady=2)

        # option menu for filtering the error type
        self._error_type_list = ["(no types)"]
        self._error_type_option_menu = tk.OptionMenu(self._gridErrors,
                                                     tk.StringVar(self._gridErrors, value="(no types)"),
                                                     *self._error_type_list)
        self._error_type_option_menu.config(relief=tk.GROOVE,
                                            bg=self.ide.theme.get_bg_color(),
                                            highlightbackground=self.ide.theme.get_bg_color(),
                                            fg=self.ide.theme.get_fg_color())
        self._error_type_option_menu.grid(row=0, column=3, sticky=tk.W + tk.E, pady=2)

        # make the infos stretch with the window
        tk.Grid.columnconfigure(self._gridErrors, 0, weight=1)
        tk.Grid.columnconfigure(self._gridErrors, 1, weight=1)
        tk.Grid.columnconfigure(self._gridErrors, 2, weight=50)
        tk.Grid.columnconfigure(self._gridErrors, 3, weight=50)
        self._frameErrors = ttk.Frame(self._gridErrors)
        self._frameErrors.grid(row=1, column=0, columnspan=4, sticky=tk.W + tk.E + tk.N + tk.S)
        tk.Grid.rowconfigure(self._gridErrors, 0, weight=1)
        tk.Grid.rowconfigure(self._gridErrors, 1, weight=100)

        self.tree_view_errors = ttk.Treeview(self._frameErrors, selectmode='browse',
                                             column=('Type', 'Message', 'Line', 'Column', 'File'))
        self.tree_view_errors.heading("#0", text="", anchor=tk.W)
        self.tree_view_errors.heading("#1", text="Code", anchor=tk.W)
        self.tree_view_errors.heading("#2", text="Message", anchor=tk.W)
        self.tree_view_errors.heading("#3", text="Line", anchor=tk.E)
        self.tree_view_errors.heading("#4", text="Column", anchor=tk.E)
        self.tree_view_errors.heading("#5", text="File", anchor=tk.CENTER)
        self.tree_view_errors.column('#0', width=40, minwidth=40, stretch=tk.NO, anchor=tk.W)
        self.tree_view_errors.column('Type', width=40, minwidth=40, stretch=tk.YES, anchor=tk.W)
        self.tree_view_errors.column('Message', width=440, minwidth=440, stretch=tk.YES, anchor=tk.W)
        self.tree_view_errors.column('Line', width=30, minwidth=30, stretch=tk.YES, anchor=tk.E)
        self.tree_view_errors.column('Column', width=30, minwidth=30, stretch=tk.YES, anchor=tk.E)
        self.tree_view_errors.column('File', width=70, minwidth=70, stretch=tk.YES, anchor=tk.CENTER)
        self.tree_view_errors.bind('<Double-Button-1>', self.__tree_view_error_clicked)
        self.__add_scrollbar(self._frameErrors, self.tree_view_errors)
        self.tree_view_errors.pack(expand=True, fill=tk.BOTH)

    def update_error_warning_counts(self):
        """
        Update the counts in the IDE depending on the error list
        :return: None
        """
        error_num = 0
        warning_num = 0
        for i in self.tree_view_errors.get_children():
            item = self.tree_view_errors.item(i)
            if type(item['image']) is list:
                if item['image'][0] == 'warning':
                    warning_num += 1
                elif item['image'][0] == 'cancel':
                    error_num += 1
        self._label_error_num.config(text="Errors (" + str(error_num) + ")")
        self._label_warning_num.config(text="Warnings (" + str(warning_num) + ")")

    def refresh_boxes(self, set_of_errors: set):
        # delete all old items in tree_view that belong to the current transformer, i.e. have its name
        self.tree_view_errors.delete(*self.tree_view_errors.get_children())
        # convert the set to list to make it sortable and sort it by the sortkey of the error
        list_of_errors = list(set_of_errors)
        list_of_errors.sort(reverse=False, key=lambda x: x.sort_key())
        # insert new items (if any) in tree_view
        for item in list_of_errors:
            if item.mainType == "E":
                im = self.ide.model.images["cancel"]
            else:
                im = self.ide.model.images["warning"]
            item_tuple = item.to_tuple()
            self.tree_view_errors.insert("", tk.END, text="", image=im, values=item_tuple)
        self.update_error_warning_counts()

    def __tree_view_error_clicked(self, event):
        """
        When the user clicks the error item, then the editor cursor will go to the position where the
        error is located.
        :param event:
        :return:
        """
        record = Utils.identify_clicked_treeview_item(self.tree_view_errors)
        if record is not None:
            if record[4] != "":
                self.ide.set_position_in_editor(record[2], record[3], record[4])
            else:
                messagebox.showerror("This is unfortunate, we are working on this!", record[1])

    def __add_scrollbar(self, frame, widget):
        scrollbar = tk.Scrollbar(frame)
        scrollbar.pack(side=tk.RIGHT, fill=tk.BOTH)
        scrollbar.config(command=widget.yview)
